bigtable/metrics.go

```diff
@@ -0,0 +1,479 @@
+			if metricDetails.metricType == otelmetric.InstrumentKindHistogram {
+				builtInMetricInstruments.distributions[metricName], err = meter.Float64Histogram(
+					metricName,
+					metric.WithDescription(metricDetails.desc),
+					metric.WithUnit(metricDetails.unit),
+				)
+				if err != nil {
+					return internalMetricsConfig, err
+				}
+			} else if metricDetails.metricType == otelmetric.InstrumentKindCounter {
+				builtInMetricInstruments.counters[metricName], err = meter.Int64Counter(
+					metricName,
+					metric.WithDescription(metricDetails.desc),
+					metric.WithUnit(metricDetails.unit),
+				)
+				if err != nil {
+					return internalMetricsConfig, err
+				}
+			}
+			internalMetricsConfig.instruments = append(internalMetricsConfig.instruments, builtInMetricInstruments)
+		}
+	}
+	return internalMetricsConfig, nil
+}
+
+func (config *metricsConfigInternal) getOperationRecorder(ctx context.Context, table string, appProfile string, method string, isStreaming bool) (*builtinMetricsTracer, func()) {
+	mt := newBuiltinMetricsTracer(method, table, appProfile, isStreaming)
+	return &mt, config.recordOperationCompletion(ctx, &mt)
+}
+
+func (config *metricsConfigInternal) gaxInvokeWithRecorder(ctx context.Context, mt *builtinMetricsTracer,
+	f func(ctx context.Context, _ gax.CallSettings) error, opts ...gax.CallOption) error {
+
+	callWrapper := func(ctx context.Context, callSettings gax.CallSettings) error {
+		// Increment number of attempts
+		mt.attemptCount++
+
+		recorder := config.recordAttemptCompletion(ctx, mt)
+		defer recorder()
+
+		err := f(ctx, callSettings)
+
+		// Record attempt status
+		statusCode, _ := convertToGrpcStatusErr(err)
+		mt.status = statusCode.String()
+		return err
+	}
+	return gax.Invoke(ctx, callWrapper, opts...)
+}
+
+// recordAttemptCompletion returns a function that should be executed to record attempt specific metrics
+// It records as many metrics as it can and does not return error
+func (config *metricsConfigInternal) recordAttemptCompletion(ctx context.Context, mt *builtinMetricsTracer) func() {
+	if !config.builtInEnabled {
+		return noOpRecordFn
+	}
+	startTime := time.Now()
+
+	return func() {
+		// Calculate elapsed time
+		elapsedTime := time.Since(startTime).Nanoseconds()
+
+		// Attributes for attempt_latencies
+		attemptLatCurrCallAttrs, attemptLatErr := mt.toOtelMetricAttrs(metricNameAttemptLatencies)
+		attemptLatAllAttrs := metric.WithAttributes(append(config.clientAttributes, attemptLatCurrCallAttrs...)...)
+
+		// Attributes for server_latencies
+		serverLatCurrCallAttrs, serverLatErr := mt.toOtelMetricAttrs(metricNameServerLatencies)
+		serverLatAllAttres := metric.WithAttributes(append(config.clientAttributes, serverLatCurrCallAttrs...)...)
+
+		for _, builtInMetricInstruments := range config.instruments {
+			if attemptLatErr == nil {
+				builtInMetricInstruments.distributions[metricNameAttemptLatencies].Record(ctx, float64(elapsedTime), attemptLatAllAttrs)
```

Comment:
[Default is cumulative](https://github.com/open-telemetry/opentelemetry-go/blob/sdk/metric/v1.27.0/sdk/metric/reader.go#L131) but updated the code to directly use cumulative since default could change.